src/Algebra/Graph/AdjacencyIntMap/Algorithm.hs

```diff
@@ -213,32 +239,86 @@ dfs vs = concatMap flatten . dfsForestFrom vs
-reachable x = concat . bfs [x]
+reachable x = dfs [x]
+
+type Cycle = NonEmpty
+type TopSort = Either (Cycle Int) [Int]
+data Entry = Entered Int | Exited Int
+type ParentTable = IntMap.IntMap Entry
+data S = S { table :: !ParentTable, order :: [Int] }
+
+topSort' :: (MonadState S m, MonadCont m) => AdjacencyIntMap -> m TopSort
+topSort' g = callCC $ \cyclic ->
+  do let vertices = map fst $ IntMap.toDescList $ adjacencyIntMap g
+         adjacent = IntSet.toDescList . flip postIntSet g
+         dfs prev curr =
+           do enter prev curr
+              forM_ (adjacent curr) $ \next ->
+                nodeState next >>= \case
+                  Nothing -> dfs curr next
+                  Just (Exited _) -> return ()
+                  _ -> cyclic . Left . retrace curr next =<< gets table
+              exit curr
+     forM_ vertices $ \v -> nodeState v >>= \case
+       Nothing -> dfs undefined v
+       _ -> return ()
+     Right <$> gets order
+  where
+    nodeState v = gets (IntMap.lookup v . table)
+    enter u v = modify' (\(S p vs) -> S (IntMap.insert v (Entered u) p) vs)
+    exit v = modify' (\(S p vs) -> S (IntMap.alter (fmap leave) v p) (v:vs))
+      where leave = \case
+              Entered x -> Exited x
+              _ -> error $ "Internal error: dfs search order violated"
+    retrace curr head table = aux (curr :| []) where
+      aux xs@(curr :| _)
+        | head == curr = xs
+        | otherwise = case table IntMap.! curr of
+            Entered prev -> aux (prev <| xs)
+            _ -> error "Internal error: dfs back edge mistakingly identified"
 
--- | Compute the /topological sort/ of a graph or return @Nothing@ if the graph
--- is cyclic.
+-- | Compute a topological sort of a DAG or discover a cycle.
+--
+--   Vertices are expanded in increasing order with respect to their
+--   'Ord' instance. This gives the lexicographically smallest
+--   topological ordering in the case of success. In the case of
+--   failure, the cycle is characterized by being the
+--   lexicographically smallest up to rotation with respect to @Ord
+--   (Dual Int)@ in the first connected component of the graph
+--   containing a cycle, where the connected components are ordered by
+--   their largest vertex with respect to @Ord a@.
+--
+--   Complexity: /O((n+m)*min(n,W))/ time and /O(n)/ space.
 --
 -- @
--- topSort (1 * 2 + 3 * 1)               == Just [3,1,2]
--- topSort (1 * 2 + 2 * 1)               == Nothing
--- fmap ('flip' 'isTopSortOf' x) (topSort x) /= Just False
--- 'isJust' . topSort                      == 'isAcyclic'
+-- topSort (1 * 2 + 3 * 1)                    == Right [3,1,2]
+-- topSort ('path' [1..5])                      == Right [1..5]
+-- topSort (3 * (1 * 4 + 2 * 5))              == Right [3,1,2,4,5]
+-- topSort (1 * 2 + 2 * 1)                    == Left (2 ':|' [1])
+-- topSort ('path' [5,4..1] + 'edge' 2 4)         == Left (4 ':|' [3,2])
+-- topSort ('circuit' [1..3])                   == Left (3 ':|' [1,2])
```

Comment:
I wish we could return `[1..3]` in this case. Is it difficult to achieve? Right now all outputs seem unintuitive and it would be great if at least some of them were.

Reply:
I could change it to scan the cycle for the minimum, and present from there. But I'm not sure that makes the result more intuitive? Currently, the first vertex listed is the first one visited during the search that belongs to the cycle

Reply:
> Currently, the first vertex listed is the first one visited during the search that belongs to the cycle

But `3` is not "the first one visited during the search that belongs to the cycle", is it? Yet it is listed first, this is what makes the result unintuitive to me. The first vertex on a cycle that we visit here is `1`, not `3`.

Reply:
It is, since the list of vertices that gets searched is descending. In the case of `circuit [1..3]`, `3` is the first node expanded in `dfs`, it is the root of the search tree. 

I went with this ordering so we get the lexicographically least topological sort when there are no cycles. Since the `topSort` is the list of vertices sorted by their exit times during `dfs`, expanding the smaller vertices as late as possible gives the lexicographically least such enumeration.

Reply:
Ah, my apologies, I keep forgetting about the fact that we use descending order.

I'm quite perplexed by the fact that it's difficult to have two good features, even though each feature seems easy enough on its own :)

Reply:
Haha yeah. I was thinking it might be nice to have a search that wouldn't exit on the first cycle, but would classify all back edges found during the search and all tree edges.

This would lose some performance for the specific case where one wants to check for any cycle at all, but could be useful more generally.

Reply:
> it might be nice to have a search that wouldn't exit on the first cycle, but would classify all back edges found during the search and all tree edges.

Hmm, sounds interesting. Could we find something like a forest of all "fundamental cycles" of a graph, and then use it compute not just one cycle, but all cycles if need be? 

It's probably not a very good idea since there will be around 0 users for this cool feature... :-)

OK, let's stop here and keep the order as is. I feel like I'm just overthinking it -- most likely, whatever cycle we return, the user will be happy :)